### DISCUSSION TITLE
/the-basics/logging.html

### DISCUSSION THREAD

**Discussion by giscus** [goravel/goravel#326] [Announcements]
# /the-basics/logging.html

[[toc]] Introduction In order to understand the running status of the application, Goravel provides a powerful log service that can record log messages and system errors to a fi...

https://www.goravel.dev/the-basics/logging.html

**Comment by kuchaguangjie** [comment]
When print log it show `local.info:`, so how to make it print code file & line number.

**Reply by hwbrzzl** [reply]
`facades.Log().Error()` will add trace info automatically.

**Reply by kuchaguangjie** [reply]
ok, but is it possible to let `Info()` to print the file name & line number too ?

**Reply by hwbrzzl** [reply]
Sure, please follow https://github.com/goravel/goravel/issues/332.

**Reply by raydenka21** [reply]
Can you send parameters to create a custom filename?

**Comment by FishyKun** [comment]
facades.Log()怎么在调用的时候自主切换通道，类似于:Log::channel('slack')->info('Something happened!');

**Reply by hwbrzzl** [reply]
抱歉，目前还不行，我新建一个 issue 记录下，感谢反馈。

**Comment by raydenka21** [comment]
Can you send parameters to create a custom filename?

**Reply by hwbrzzl** [reply]
For example?

**Comment by incodemode** [comment]
Is there a way to add an id to every request so lets say some does /users so the rand id like 640ddd60
facades.Log().Info("listing users\ntest")
The log would look like this:
[2024-12-05 03:19:18] | 640ddd60 | st1.info: listing users
640ddd60 | test
[HTTP] 2024/12/04 - 21:19:18 | 640ddd60 | 200 |      1.5352ms |             ::1 | GET      "/health"

I was able to create the extension and custom Logger even a middleware to create the ctx requestId and the a context so I don't have to be passing the context but I have not been able to avoid logrus from writing the log twice. Any advice on duplicated logs?

**Reply by hwbrzzl** [reply]
`duplicated logs`? What's the meaning, please?

**Reply by incodemode** [reply]
I copied the custom logger and added my own code but when I do the print:
    facades.Log().Infof("%s\n%s", "test1", "test2")
The following is printed:
    [8123a28a] context=context.Background level=info time=2024-12-05 19:42:57.9305314 -0600 CST m=+143.212008401 message=test1
    test2
    [2024-12-06 01:42:57] st1.info: test1   //repeated!!
    test2
this is my CustomLogger
    package extensions

    import (
        "context"
        "fmt"
        "goravel/app/http/middleware"
        "goravel/app/logctx"

        "github.com/goravel/framework/contracts/log"
    )

    type CustomLogger struct {
    }

    // Handle pass channel config path here
    func (logger *CustomLogger) Handle(channel string) (log.Hook, error) {
        return &Hook{}, nil
    }

    type Hook struct {
    }

    // Levels monitoring level
    func (h *Hook) Levels() []log.Level {
        return []log.Level{
            log.DebugLevel,
            log.InfoLevel,
            log.WarningLevel,
            log.ErrorLevel,
            log.FatalLevel,
            log.PanicLevel,
        }
    }

    // Fire execute logic when trigger
    func (h *Hook) Fire(entry log.Entry) error {
        //requestID := getRequestID(entry.Context())
        ctx := logctx.GetContext()

        // Extract request-specific information (e.g., request ID) from the context
        requestID := "unknown"
        if id, ok := ctx.Value("request_id").(string); ok {
            requestID = id
        }

        fmt.Println(fmt.Sprintf("[%s] context=%v level=%v time=%v message=%s", requestID, entry.Context(), entry.Level(), entry.Time(), entry.Message()))
        return nil
    }

I config it like this in config/logging.go
    "custom": map[string]interface{}{
        "driver": "custom",
        "via":    &extensions.CustomLogger{},
    },
and the default is custom.

**Reply by hwbrzzl** [reply]
Thanks, I got you, I'll fix it recently.

**Reply by hwbrzzl** [reply]
Created an issue: https://github.com/goravel/goravel/issues/531

**Reply by hwbrzzl** [reply]
Solved in goravel/framework v1.14.9
